### PR TITLE
cmd/node: configuration changes for genesis id

### LINKED ISSUE #3505: cmd/node: configuration changes for genesis id
part of: https://github.com/spacemeshos/pm/issues/135
depends on: https://github.com/spacemeshos/go-spacemesh/issues/3505

- add a genesis section into config
- for this issue genesis section needs to include: genesis time and genesis extra data (other fields can be moved into this config later)
   - genesis time is a string with time encoded according to RFC3339
   - genesis extra data is a string with a max size of 255 chars. defaults to `mainnet`
- when node is started check if genesis section is already stored in the data directory (example: /data/genesis.conf)
   - if it is not stored, encode genesis section with json and persist it on disk
   - if stored - decode object with json into another instance of genesis config, compute diff using https://github.com/google/go-cmp/ and print fatal error to the operator with a diff and a path to persisted genesis conf
- compute genesis id as a hash from genesis time and genesis extra data
- golden atx should be set to genesis id

### PR DESCRIPTION
**Pull request by evt**
## Motivation
Closes #3505 

## Changes
Adds a genesis section into config.

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: config/genesis.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		ExtraData:   DefaultTestGenesisExtraData(),
+		GoldenATXID: DefaultTestGoldenATXId().Hash32().Hex(),
+		GenesisID:   CalcGenesisId([]byte(DefaultTestGenesisExtraData()), DefaultTestGenesisTime()).Hex(),
+		Path:        DefaultGenesisDataDir(),
+	}
+}
+
+func (gc *GenesisConfig) Compare(payload *GenesisConfig) error {
+	if diff := cmp.Diff(payload, gc); diff != "" {
+		fmt.Printf("config files mismatch (-want +got):\n%s", diff)
```
Comment (countvonzero) [inline config/genesis.go]: rm

File: cmd/node/node.go [hunk 2]
```diff
@@ -1164,3 +1170,35 @@ func decodeLoggers(cfg config.LoggerConfig) (map[string]string, error) {
...
+			return fmt.Errorf("failed to read stored genesis config file: %s", err)
+		}
+
+		var payload config.GenesisConfig
+
+		if err = json.Unmarshal(genesisData, &payload); err != nil {
+			return fmt.Errorf("failed to unmarshal stored genesis config file: %w", err)
+		}
+		if err = app.Config.Genesis.Compare(&payload); err != nil {
+			return fmt.Errorf("failed to compare stored genesis config file: %w", err)
```
Comment (countvonzero) [inline cmd/node/node.go]: the txt is misleading. how about "genesis config changed from previous run"?

File: cmd/node/node.go [hunk 3]
```diff
@@ -1164,3 +1170,35 @@ func decodeLoggers(cfg config.LoggerConfig) (map[string]string, error) {
...
+		}
+		return nil
+	}
+	jsonCfg, err := json.Marshal(app.Config.Genesis)
```
Comment (countvonzero) [inline cmd/node/node.go:1192]: only persist when the genesis config is absent?

**Review by countvonzero — COMMENTED** [review]
i don't know the spec well enough to approve. just some general comment

File: cmd/node/node.go [hunk 4]
```diff
@@ -1164,3 +1170,35 @@ func decodeLoggers(cfg config.LoggerConfig) (map[string]string, error) {
 	}
 	return rst, nil
 }
+
+func (app *App) checkAndStoreGenesisConfig() error {
+	genesisConfigPath := filepath.Join(app.Config.Genesis.Path, config.DefaultGenesisConfigFileName)
```
Comment (dshulyak) [inline cmd/node/node.go]: this should use configured data dir, not some new parameter

File: cmd/node/node.go [hunk 5]
```diff
@@ -1164,3 +1170,35 @@ func decodeLoggers(cfg config.LoggerConfig) (map[string]string, error) {
...
+	if exists := filesystem.PathExists(genesisConfigPath); exists {
+		genesisData, err := ioutil.ReadFile(genesisConfigPath)
+		if err != nil {
+			return fmt.Errorf("failed to read stored genesis config file: %s", err)
+		}
+
+		var payload config.GenesisConfig
```
Comment (dshulyak) [inline cmd/node/node.go]: payload is not a good name

File: cmd/node/node.go (hunk 2 above)
Comment (dshulyak) [reply cmd/node/node.go]: yes message is bad. it also needs to specify location of persisted config file

File: config/genesis.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	DefaultGenesisConfigFileName = "genesis.conf"
+)
+
+type GenesisConfig struct {
+	Accounts    *apiConfig.GenesisAccountConfig `mapstructure:"genesis-api"`
+	GenesisTime string                          `mapstructure:"genesis-time"`
+	ExtraData   string                          `mapstructure:"genesis-extradata"`
+	GoldenATXID string
+	GenesisID   string
+	Path        string
```
Comment (dshulyak) [inline config/genesis.go]: this config should have only 3 things: Accounts (mapping), GenesisTime, ExtraData

remove everything else

File: config/genesis.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+func (gc *GenesisConfig) Compare(payload *GenesisConfig) error {
+	if diff := cmp.Diff(payload, gc); diff != "" {
+		fmt.Printf("config files mismatch (-want +got):\n%s", diff)
+		return errors.New("failed to match config files")
+	}
+	return nil
+}
+
+func DefaultGenesisDataDir() string {
```
Comment (dshulyak) [inline config/genesis.go]: remove

File: config/genesis.go [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+	return filepath.Join(defaultDataDir, DefaultGenesisConfigFileDir)
+}
+
+func DefaultGenesisTime() string {
+	return time.Now().Format(time.RFC3339)
+}
+
+func DefaultGenesisExtraData() string {
```
Comment (dshulyak) [inline config/genesis.go]: put this line into the default config or to constant

File: config/genesis.go [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+	return "mainnet"
+}
+
+func DefaultGoldenATXId() types.ATXID {
```
Comment (dshulyak) [inline config/genesis.go]: remove, it doesn't make sense

File: config/genesis.go [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+func DefaultTestnetGenesisExtraData() string {
+	return "testnet"
+}
+
+func DefaultTestnetGoldenATXId() types.ATXID {
+	return types.ATXID(CalcGoldenATX([]byte(DefaultTestnetGenesisExtraData()), DefaultTestnetGenesisTime()))
+}
+
+func CalcGenesisId(genesisExtraData []byte, genesisTime string) types.Hash20 {
```
Comment (dshulyak) [inline config/genesis.go]: this function is not doing a right thing, it should be

hash.Write(gentime)
hash.Write(extradata)

File: config/genesis.go [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+	}
+	digest := hasher.Sum([]byte{})
+	return types.BytesToHash(digest).ToHash20()
+}
+
+func (cfg *GenesisConfig) CalcGenesisID() {
+	cfg.GenesisID = CalcGenesisId([]byte(cfg.ExtraData), cfg.GenesisTime).Hex()
+}
+
+func CalcGoldenATX(genesisData []byte, genesisTime string) types.Hash32 {
```
Comment (dshulyak) [inline config/genesis.go]: remove this

File: config/genesis.go [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+		copy(padded[0:dataLen], genesisExtraData)
+		hasher.Write(padded)
+	} else {
+		hasher.Write(genesisExtraData)
+	}
+	digest := hasher.Sum([]byte{})
+	return types.BytesToHash(digest).ToHash20()
+}
+
+func (cfg *GenesisConfig) CalcGenesisID() {
```
Comment (dshulyak) [inline config/genesis.go]: return types.Hash20 from this method

File: config/genesis.go [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+func DefaultTestnetGenesisTime() string {
+	return DefaultTestGenesisTime()
+}
+
+func DefaultTestnetGenesisExtraData() string {
+	return "testnet"
+}
+
+func DefaultTestnetGoldenATXId() types.ATXID {
```
Comment (dshulyak) [inline config/genesis.go]: remove

File: config/genesis.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+)
+
+const (
+	ExtraDataLen                 = 255
+	DefaultGenesisConfigFileDir  = "data"
+	DefaultGenesisConfigFileName = "genesis.conf"
+)
+
+type GenesisConfig struct {
+	Accounts    *apiConfig.GenesisAccountConfig `mapstructure:"genesis-api"`
```
Comment (dshulyak) [inline config/genesis.go]: store accounts directly in this config

File: config/genesis.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+func (cfg *GenesisConfig) CalcGenesisID() {
+	cfg.GenesisID = CalcGenesisId([]byte(cfg.ExtraData), cfg.GenesisTime).Hex()
+}
+
+func CalcGoldenATX(genesisData []byte, genesisTime string) types.Hash32 {
+	return CalcGenesisId([]byte(genesisData), genesisTime).ToHash32()
+}
+
+func (cfg *GenesisConfig) CalcGoldenATX() {
```
Comment (dshulyak) [inline config/genesis.go]: remove this

**Review by dshulyak — CHANGES_REQUESTED** [review]
changes in the vm module needs to be removed
also please drop all that DefaultSomethingSomething. please keep only DefaultConfig. DefaultTestnetConfig and DefaultFastnetConfig should be written in presets files

File: config/genesis.go (hunk 1 above)
Comment (evt) [reply config/genesis.go]: I added this on purpose to see the diff in a clean way when error occurs. If we include it into error printed then JSON log escapes it and it doesn't look good to understand. Or no point to print diff at all?

File: cmd/node/node.go (hunk 3 above)
Comment (evt) [reply cmd/node/node.go:1192]: yes, correct. If it exists - we make sure it hasn't changed and exit otherwise.

File: config/genesis.go (hunk 10 above)
Comment (evt) [reply config/genesis.go]: this is what it does, not sure what you mean

Comment (dshulyak) [reply config/genesis.go]: remove padding

File: cmd/node/node.go (hunk 5 above)
Comment (evt) [reply cmd/node/node.go]: fixed

File: config/genesis.go (hunk 10 above)
Comment (evt) [reply config/genesis.go]: done

File: config/genesis.go (hunk 15 above)
Comment (evt) [reply config/genesis.go]: done

File: cmd/node/node.go (hunk 2 above)
Comment (evt) [reply cmd/node/node.go]: fixed

File: cmd/node/node.go (hunk 4 above)
Comment (evt) [reply cmd/node/node.go]: fixed

File: config/genesis.go (hunk 11 above)
Comment (evt) [reply config/genesis.go]: removed

File: config/genesis.go (hunk 12 above)
Comment (evt) [reply config/genesis.go]: removed as GenesisID removed

File: config/genesis.go (hunk 6 above)
Comment (evt) [reply config/genesis.go]: done

File: config/genesis.go (hunk 13 above)
Comment (evt) [reply config/genesis.go]: done

File: config/genesis.go (hunk 8 above)
Comment (evt) [reply config/genesis.go]: done

File: config/genesis.go (hunk 9 above)
Comment (evt) [reply config/genesis.go]: removed

File: config/genesis.go (hunk 7 above)
Comment (evt) [reply config/genesis.go]: done

File: config/genesis.go (hunk 14 above)
Comment (evt) [reply config/genesis.go]: done

**Comment by evt** [comment]
> changes in the vm module needs to be removed also please drop all that DefaultSomethingSomething. please keep only DefaultConfig. DefaultTestnetConfig and DefaultFastnetConfig should be written in presets files

vm changes removed. DefaultTestnetConfig - done. DefaultFastnetConfig - do we have it?

**Comment by evt** [comment]
bors try

File: config/genesis.go [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+			Address: genesisAddr,
+			Balance: balance,
+		})
+	}
+	return rst
+}
+
+func (gc *GenesisConfig) Compare(stored *GenesisConfig, path string) error {
+	if diff := cmp.Diff(stored, gc); diff != "" {
+		fmt.Printf("Genesis config %s changed from previous run (-want +got):\n%s", path, diff)
```
Comment (dshulyak) [inline config/genesis.go]: but this can't stay like this, can you find a cleaner solution to print details?

File: config/genesis.go [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/spacemeshos/go-spacemesh/hash"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+const (
+	DefaultGenesisConfigFileDir  = "data"
+	DefaultGenesisConfigFileName = "genesis.conf"
+	defaultGenesisExtraData      = "mainnet"
+
+	DefaultTestGenesisTime      = "2022-12-25T00:00:00+00:00"
```
Comment (dshulyak) [inline config/genesis.go]: where do we use this? should be possible to cleanup

File: config/genesis.go [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/google/go-cmp/cmp"
+
+	apiConfig "github.com/spacemeshos/go-spacemesh/api/config"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/hash"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+const (
+	DefaultGenesisConfigFileDir  = "data"
```
Comment (dshulyak) [inline config/genesis.go]: so given that spacemesh is configured with datadir - /state/ . genesis config will saved into /state/data/genesis.conf. does it make sense? why is data needed?

File: config/genesis.go [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+func defaultGenesisConfig() *GenesisConfig {
+	return &GenesisConfig{
+		Accounts:    apiConfig.DefaultGenesisAccounts(),
+		GenesisTime: time.Now().Format(time.RFC3339),
+		ExtraData:   defaultGenesisExtraData,
+	}
+}
+
+func GenesisDataDir(dataDirParent string) string {
```
Comment (dshulyak) [inline config/genesis.go]: can you explain why do we need this?

File: api/config/genesis.go [hunk 20]
```diff
@@ -3,51 +3,23 @@ package config
 import (
 	"github.com/spacemeshos/go-spacemesh/common/types"
 	"github.com/spacemeshos/go-spacemesh/common/util"
-	"github.com/spacemeshos/go-spacemesh/log"
 	"github.com/spacemeshos/go-spacemesh/signing"
 )
 
-// GenesisConfig defines accounts that will exist in state at genesis.
```
Comment (dshulyak) [inline api/config/genesis.go:10]: why other things weren't moved?

File: config/genesis.go (hunk 18 above)
Comment (evt) [reply config/genesis.go]: Not needed really. Removed.

File: config/genesis.go (hunk 19 above)
Comment (evt) [reply config/genesis.go]: Removed along with https://github.com/spacemeshos/go-spacemesh/pull/3568#discussion_r977193346

File: config/genesis.go (hunk 16 above)
Comment (evt) [reply config/genesis.go]: replaced with log error

File: api/config/genesis.go (hunk 20 above)
Comment (evt) [reply api/config/genesis.go:10]: moved it all to `config/genesis.go`

File: config/genesis.go (hunk 17 above)
Comment (evt) [reply config/genesis.go]: moved it to `cmd/node/node_test.go`

File: config/genesis.go [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+		}
+		rst = append(rst, types.Account{
+			Address: genesisAddr,
+			Balance: balance,
+		})
+	}
+	return rst
+}
+
+func (gc *GenesisConfig) Compare(stored *GenesisConfig, path string, lg log.Log) error {
```
Comment (dshulyak) [inline config/genesis.go:73]: this method signature is very bad

`Diff(other genesisConfig) string`

if result is not empty print a fatal error on the caller side

**Comment by dshulyak** [comment]
implemented elsewhere

**Pull request closed without merging**